Make MA work on Pharo 9/10 and on GT

janvrany opened the pull request.

janvrany committed:
Extract currying overrides into version-specific packages

This commit creates new packages, one for Pharo8 and another for
Pharo9 an newer that contain overrides for `BlockClosure >> value:...`.

This is necessary since primitive numbers have changed from one Pharo
version to another, causing the VM to crash when using wrong primitive
number.

With this commit, MachineArithmetic loads into Pharo 10.

janvrany committed:
Load external `#mustBeBoolean` deoptimization support package on P10

Pharo 10 removed built-in support for #mustBeBoolean deoptimization
(see [1]). This commit loads an external package [2] to bring this
essential functionality back.

[1]: https://github.com/pharo-project/pharo/commit/94cff3ad2fbd38eeb9231e4dd5c2894379f2849d
[2]: https://github.com/pharo/MustBeBooleanMagic

janvrany committed:
Rewrite `Z3CAPITest >> #testFindModel0` without using compiler API

Compiler API apparently changed between P9 and P10, making this test failing on
P10. There's no real need to go through compiler in this test, so this commit changes
the code to not using it. 

As a side-effect, this makes this test working on Smalltalk/X too.

janvrany committed:
Add P10-specific methods to `CardanoTartagliaContext`

This commit adds some `Context` methods introduced after P8 to 
`CardanoTartagliaContext`. With this new methods, Refinements-Tests
all pass on P10.

janvrany committed:
CI: add P10, P11, P12 and P13 to test matrix as experimental

janvrany committed:
Use (new) `#mustBeBooleanMagicDoitForContext:` instead of  `#asDoitForContext:`

...since the latter was removed from P11. Similarly, use (new)
`#rewriteTempsForMustBeBooleanMagicContext:` instead of `#rewriteTempsForContext:`.

janvrany committed:
Add P11-specific methods to `CardanoTartagliaContext`

In P11, the Opal Compiler changes quite a bit so this commit add more methods
to `CardanoTartagliaContext` to make it work also on P11. This required introducing
another helper class `CardanoTartagliaContextVariable` to make Opal compiler happy.

With these changes, `Refinements-Tests` all pass on P11.

janvrany commented:
FYI @shingarov: with latest version of this PR, MA tests pass on P8, P10 and P11 on both, Linux and Windows. 
See https://github.com/shingarov/MachineArithmetic/pull/52/checks

I tried P9 but there's something weird and given that current "stable" Pharo is P11 and P12 already on the way, I won't bother.

shingarov commented:
I am looking at the latest commits and I am not even sure what is worse in terms of maintenance overhead: fixing P8 (possibly compiling our own VM to not break libcurl), or fixing P11.  If a Smalltalk-80 dialect has broken `#ifTrue:`, MachineArithmetic-on-that-dialect will wait until it gets fixed.

janvrany commented:
Fine with me, but that effectively means we stay on P8 until **we** fix it elsewhere : - )

shingarov commented:
One way or another, that `#ifTrue:` is NOT part of MA, that much is for sure.  So perhaps — if we really REALLY wanted P11 — we could load into a "Usable P11", which could be built including a tag inside our own fork of MustBeBooleanMagic.  But to drag that insanity *into* MA, I just don't see how this is any better than the problems with P8.

janvrany commented:
If that's one only concern, I can certainly move 16f92eb and 5146f3c to https://github.com/janvrany/pharo-hacks (Usable ... might not be a good place).

shingarov commented:
The concern is that the sum of P11 hacks is becoming a bigger mess than whatever maintenance overhead would be required to stay on P8.

shingarov commented:
In other words: P11 is *more* broken than P8.

janvrany commented:
However, for now IMO the major blocker here is removal of GT inspector, see https://github.com/pharo-project/pharo/commit/0fdf1ae3248f93eea90adbfbf29ece53f933e360

shingarov commented:
Yeah that's another one.  In summary: P11 is a non-starter.

janvrany committed:
Move copy of MustBeBooleanMagic package to MachineArithmetic

This commit add a copy of MustBeBooleanMagic ([1], commit c5ae7bc)
to MachineArithmetic. This is to allow for further modifications
to make it work also on Pharo 11 and later.

Another option would be to fork it and then maintain yet another fork,
but for such a small (and likely abandoned) package this seems to too
much maintenance overhead and we already have "repository plaque".
